torchtune/modules/attention.py

```diff
@@ -221,7 +226,7 @@ def forward(
             v,
             attn_mask=mask,
             dropout_p=self.attn_dropout,
-            is_causal=self.kv_cache is None and mask is None,
+            is_causal=self.kv_cache is None and mask is None and self.is_causal,
```

Comment:
It will be confusing to the user if they set is_causal and it doesn't end up actually setting this value to true because one of the other conditions isn't true. We should either validate these inputs and raise if for example they set is_causal=True but pass in a mask, or just remove the is_causal parameter altogether.

Reply:
The complex part of the code is GroupedQueryAttention. There was nothing inherently causal about it or nothing that made it self attention after the definition of q, k, v. We could split GQA into a self attention and cross attention variant but then all of the complex parts would get duplicated and only a few parts of the interface would change. That's why I moved that up to the transformer layer, but layers need GQA and they know which are the correct inputs to provide to it.

On is_causal, what this flag is really address is "what is my default mask when none is provided". During cross attention, caching, and sample packing, we provide a mask and it just uses that. But when none is provided we are currently defaulting to a causal mask. This is great in decoders but not what we want our default behavior to be in encoder self attention layers.

One proposal would be naming this "default_mask_is_causal" or "causal_default_mask"